lib/adapters/parsers/grok.js

```diff
@@ -0,0 +1,76 @@
+var base = require('./base');
+var errors = require('../../errors');
+var grok = require('node-grok');
+var Promise = require('bluebird');
+var readlines = require('readlines');
+
+module.exports = base.extend({
+
+    initialize: function(options) {
+        this.pattern = options.pattern || '%{GREEDYDATA:message}';
+    },
+
+    parseStream: function(stream, emit) {
+        var self = this;
+
+        return new Promise(function(resolve, reject) {
+            grok.loadDefault(function(patterns) {
+                var buffer = [];
+                var liner = readlines.readlineStream();
+                var pattern = patterns.createPattern(self.pattern);
+
+                // because of the asynchronous way that patterns parse we need
+                // to track the pending parses and when the stream has ended to
+                // do the right thing.
+                var pending = 0;
+                var streamEnd = false;
+
+                liner.on('readable', function(){
+                    var line;
+                    while ((line = liner.read())) {
+                        line = line.toString();
+                        pending++;
+                        pattern.parse(line, function (err, point) {
+                            if (err) {
+                                reject(errors.runtimeError('RT-INTERNAL-ERROR',{
```

Comment:
So now that I'm fixing this I've figured out that the asynchronous parsing is leading to some really buggy code since with grok.parse being asynchronous we actually have no gaurantee on the order of the calling back... thinking about how to correctly fix this.